File: src/main/java/openmedstack/Chassis.java
package openmedstack;

import com.google.inject.Module;
import openmedstack.commands.CommandResponse;
import openmedstack.commands.DomainCommand;
import openmedstack.events.BaseEvent;

import java.util.HashMap;
import java.util.concurrent.CompletableFuture;
import java.util.function.BiConsumer;
import java.util.function.Consumer;
import java.util.function.Function;

public class Chassis implements AutoCloseable {
    private final DeploymentConfiguration _configuration;
    private final HashMap<String, Object> _metadata;
    private Function<DeploymentConfiguration, Service> _builder = Chassis::buildService;
    private Service _service;

    private Chassis(DeploymentConfiguration configuration) {
        _configuration = configuration;
        _metadata = new HashMap<>();
    }

    public DeploymentConfiguration getConfiguration() {
        return _configuration;
    }

    public static Chassis from(DeploymentConfiguration manifest) {
        return new Chassis(manifest);
    }

    public Chassis usingCustomBuilder(Function<DeploymentConfiguration, Service> builder) {
        _builder = builder;
        return this;
    }

    public CompletableFuture start() {
        _service = _builder.apply(_configuration);
        return _service.start();
    }

    public <T extends DomainCommand> CompletableFuture<CommandResponse> send(T command) throws NullPointerException {
        if (_service == null) {
            throw new NullPointerException("Chassis not started");
        }
        return _service.send(command);
    }

    public <T extends BaseEvent> CompletableFuture publish(T msg) throws NullPointerException {
        if (_service == null) {
            throw new NullPointerException("Chassis not started");
        }
        return _service.publish(msg);
    }

    public <T extends Object> T resolve(Class<T> type) throws NullPointerException {
        if (_service == null) {
            throw new NullPointerException("Chassis not started");
        }
        return _service.resolve(type);
    }

    private static Service buildService(DeploymentConfiguration manifest, Module... modules) throws IllegalStateException {
        return new DefaultService(modules);
    }

    @Override
    public void close() throws Exception {
        if(_service != null){
            _service.close();
        }
        _metadata.clear();
    }
}

File: modules/openmedstack/src/main/kotlin/org/openmedstack/Chassis.kt
package org.openmedstack

import org.openmedstack.commands.CommandResponse
import org.openmedstack.commands.DomainCommand
import org.openmedstack.events.BaseEvent
import java.util.concurrent.CompletableFuture

class Chassis private constructor(val configuration: DeploymentConfiguration) : AutoCloseable {
    private val _metadata: HashMap<String, Any> = HashMap()
    private val _packages: ArrayList<Package> = ArrayList()
    private var _builder: ((DeploymentConfiguration, Array<Package>) -> Service)? = null
    private var _service: Service? = null

    fun start(): CompletableFuture<*> {
        if (_service != null) {
            return CompletableFuture.completedFuture(true)
        }
        _service = _builder!!.invoke(configuration, _packages.toTypedArray())
        return _service!!.start()
    }

    @Throws(NullPointerException::class)
    fun <T : DomainCommand> send(command: T): CompletableFuture<CommandResponse> {
        if (_service == null) {
            throw NullPointerException("Chassis not started")
        }
        return _service!!.send(command)
    }

    @Throws(NullPointerException::class)
    fun <T : BaseEvent> publish(msg: T): CompletableFuture<*> {
        if (_service == null) {
            throw NullPointerException("Chassis not started")
        }
        return _service!!.publish(msg)
    }

    @Throws(NullPointerException::class)
    fun <T> resolve(type: Class<T>): T where T : Any {
        if (_service == null) {
            throw NullPointerException("Chassis not started")
        }
        return _service!!.resolve(type)
    }

    fun withServiceBuilder(builder: (DeploymentConfiguration, Array<Package>) -> Service): Chassis {
        _builder = builder
        return this
    }

    fun definedIn(vararg packages: Package): Chassis {
        _packages.addAll(packages)
        return this
    }

    @Throws(Exception::class)
    override fun close() {
        if (_service != null) {
            _service!!.close()
        }
        _metadata.clear()
    }

    companion object {
        fun from(manifest: DeploymentConfiguration): Chassis {
            return Chassis(manifest)
        }
    }
}
